Cortex_m backend: Add softmax support for int8 via CMSIS NN (dim=-1)

xingguo01 opened the pull request:
- integrate CMSIS softmax into Cortex-M backend
- add fusion pass/tests for quantized softmax

cc @freddan80 @per @zingo @oscarandersson8218 @digantdesai @Sebastian-Larsson @AdrianLundell @psiddh

xingguo01 commented:
@psiddh much appreciated if you can have a look

psiddh commented on backends/cortex_m/ops/op_softmax.cpp:
```diff
@@ -0,0 +1,148 @@
+    return out;
+  }
+
+  arm_softmax_s8(
```
Checking for status ?  something like

arm_status status = arm_softmax_s8(
    input_data,
    num_rows,
    row_size,
    input_multiplier_val,
    input_shift_val,
    diff_min_val,
    output_data);
if (status != ARM_MATH_SUCCESS) {
  ET_LOG(Error, "softmax_out: CMSIS NN softmax failed (status=%d)", status);
  context.fail(Error::InvalidArgument);
  return out;
}

xingguo01 replied:
Just to be clear here: https://github.com/ARM-software/CMSIS-NN/blob/main/Include/arm_nnfunctions.h#L2307. Thanks for approving the PR, my first one to Excutorch

psiddh replied:
This is great ! Thanks for your contribution ! 
Hopefully many more contributions to come :)

psiddh commented on backends/cortex_m/ops/op_softmax.cpp:
```diff
@@ -0,0 +1,148 @@
+  (void)input_zp_val; // Zero-point difference cancels out during subtraction.
+
+  validate_single_quant_params(
+      Scalar(input_zp_val),
```
Should this be 'output_zp_val' instead ?

psiddh reviewed (commented):
lgtm, left few comments

rascani commented on backends/cortex_m/ops/op_softmax.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  const int32_t input_zp_val = static_cast<int32_t>(input_zero_point);
+  const int32_t output_zp_val = static_cast<int32_t>(output_zero_point);
+  (void)input_zp_val; // Zero-point difference cancels out during subtraction.
```
Since we only need the input multiplier and shift for `arm_softmax_s8`, is there any value in including the zero points in the signature? We could probably do the output zero point value == -128 check during AOT.

xingguo01 committed:
Add softmax support for int8 in Cortex M (dim=-1)

- integrate CMSIS softmax into Cortex-M backend
- add fusion pass/tests for quantized softmax
- lint cleanup passes
- Resolved merge conflicts

Change-Id: I0ec19f011069fa1482e2de2ab62b9e7d7f56b2a8
Signed-off-by: Xingguo Li <xingguo.li@arm.com>

psiddh reviewed (approved):
lgtm
